### ISSUE TITLE
[ARCHETYPE-406] Support of velocity expressions for user-defined properties

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-archetype#393]
**[Bue Pierre-Christophe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pcbue)** opened **[ARCHETYPE-406](https://issues.apache.org/jira/browse/ARCHETYPE-406?redirect=false)** and commented

Standard properties (artifactId, groupId, ...) are supported in Velocity expressions, but still does not work with user defined properties. For example:

```xml
<requiredProperty key="a">
<defaultValue>${artifactId}.substring(0,1).toUpperCase()</defaultValue>
</requiredProperty>
```

with artifactId=toto will give a=T, but

```xml
<requiredProperty key="b"/>
<requiredProperty key="a">
<defaultValue>${b}.substring(0,1).toUpperCase()</defaultValue>
</requiredProperty>
```

with b=toto will give a=${b}.substring(0,1).toUpperCase()

---

**Affects:** backlog

**Attachments:**
- [ARCHETYPE-406.patch](https://issues.apache.org/jira/secure/attachment/12713762/ARCHETYPE-406.patch) (_3.01 kB_)
- [sample-archetype406.tar.gz](https://issues.apache.org/jira/secure/attachment/12713738/sample-archetype406.tar.gz) (_1.31 kB_)

**Issue Links:**
- [ARCHETYPE-383](https://issues.apache.org/jira/browse/ARCHETYPE-383) Use velocity expressions in descriptor properties

- [MNG-5347](https://issues.apache.org/jira/browse/MNG-5347) Support expression language in ${...}

**Remote Links:**
- [GitHub Pull Request #12
](https://github.com/apache/maven-archetype/pull/12)
- [GitHub Pull Request #16
](https://github.com/apache/maven-archetype/pull/16)
- [GitHub Pull Request #16
](https://github.com/apache/maven-archetype/pull/16)
- [GitHub Pull Request #64
](https://github.com/apache/maven-archetype/pull/64)

5 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Bue Pierre-Christophe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pcbue)** commented

ARCHETYPE-383 solves the problem for standard properties (artifactId, groupId, ...) but not for user-defined

**Comment by jira-importer** [comment]
**[Adam Israel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aisrael)** commented

Does [ARCHETYPE-397](https://jira.codehaus.org/browse/ARCHETYPE-397) address this, too?

**Comment by jira-importer** [comment]
**[Bue Pierre-Christophe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pcbue)** commented

Not exactly.

ACRHETYPE-397 deals with a problem in the display on the use of standard properties, such as artifactId.substring, and so on, but they are correctly evaluate by Velocity in the generation step of the archetype.

This bug wants to manage correctly the use of user-defined properties in the same way standard properties are, such as the description of this bug spotlights.

**Comment by jira-importer** [comment]
**[Bue Pierre-Christophe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pcbue)** commented

I can bring more precision.

I have attached an example of the issue. When user-defined propertes are specified, only the first is correctly evaluated (i think it is, but i'm not quite sure ...).

In the example, i ask for two properties : propA and propB. i capitalize these two properties, and use it in two ways : inner use (on text filtering, with $propA for example), and file name filtering. On both, capitalizePropA is correctly evaluated, but capitalizePropB is neither evaluated as propA with capitalizePropA.

Here is the interesting part of the archetype generation log :

Define value for property 'capitalizePropA':  ${propA.substring(0,1).toUpperCase()}${propA.substring(1)}: :
Define value for property 'capitalizePropB':  ${propB.substring(0,1).toUpperCase()}${propB.substring(1)}: :
Define value for property 'propA': : valA
Define value for property 'propB': : valB
Confirm properties configuration:
groupId: group
artifactId: art
version: 1.0.0-SNAPSHOT
package: group
capitalizePropA: ${propA.substring(0,1).toUpperCase()}${propA.substring(1)}
capitalizePropB: ${propB.substring(0,1).toUpperCase()}${propB.substring(1)}
propA: valA
propB: valB

For the moment, all is ok. But velocity complains on the resolution of capitalizePropB on effective generation :

[INFO] ----------------------------------------------------------------------------
[INFO] Using following parameters for creating project from Archetype: test:1
[INFO] ----------------------------------------------------------------------------
[INFO] Parameter: groupId, Value: group
[INFO] Parameter: artifactId, Value: art
[INFO] Parameter: version, Value: 1.0.0-SNAPSHOT
[INFO] Parameter: package, Value: group
[INFO] Parameter: packageInPathFormat, Value: group
[INFO] Parameter: propA, Value: valA
30 avr. 2012 12:03:34 org.apache.velocity.runtime.log.JdkLogChute log
INFO: FileResourceLoader : adding path '.'
30 avr. 2012 12:03:34 org.apache.velocity.runtime.log.JdkLogChute log
INFO: Null reference [template 'capitalizePropB', line 1, column 1] : ${propB.substring(0,1).toUpperCase()} cannot be resolved.
30 avr. 2012 12:03:34 org.apache.velocity.runtime.log.JdkLogChute log
INFO: Null reference [template 'capitalizePropB', line 1, column 38] : ${propB.substring(1)} cannot be resolved.
[INFO] Parameter: capitalizePropB, Value: ${propB.substring(0,1).toUpperCase()}${propB.substring(1)}
[INFO] Parameter: capitalizePropA, Value: ValA
[INFO] Parameter: package, Value: group
[INFO] Parameter: version, Value: 1.0.0-SNAPSHOT
[INFO] Parameter: groupId, Value: group
[INFO] Parameter: propB, Value: valB
[INFO] Parameter: artifactId, Value: art
[INFO] project created from Archetype in dir: /home/pcbue/test/plop/art
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS

**Comment by jira-importer** [comment]
**[Mickael Istria](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mickael.istria)** commented

MNG-5347 suggest allowing an expression language for user properties in $&#123;...}

**Comment by jira-importer** [comment]
**[Mark Jeffcoat](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mjeffcoat)** commented

I've attached a proposed fix. I generated the patch from the trunk, but it should also apply cleanly to 2.2.

The reason this didn't work (consistently) before is that the order in which it attempted to resolve properties was arbitrary; if a user-defined property was already in the Context, it could be used by later Velocity expressions; if it wasn't already in the Context, any reference to it would be interpreted literally. The arbitrary order of evaluation is why the examples in the comments are so confusing.

This patch forces the archetype generator to keep trying to resolve things that look like Velocity expressions until it can no longer make any further progress. It Works For Me, but I'd appreciate any comments.
